[BUG] InnerHIt result not containing a _index is throwing Hit.Index missing

### What is the bug?
_A clear and concise description of the bug._
```
Caused by: org.opensearch.client.util.MissingRequiredPropertyException: Missing required property 'Hit.index'
```
I have a query that I am getting back from opensearch that has a Hit of a parent, and alongside that, I added the option to retrieve all the children of that parent document as Innerhits. 

I've included an example below of the query, where I think it should be getting a parent that has a child document of type `childType`, and then in the innerhits, get the children for it.
```
Query.of {
    it.hasChild {
        it.type("childType")
        it.query {
            it.matchAll { it -> it }
        }
        it.innerHits {
            it.source {
                it.fetch(true)
            }
        }
    }
}
```
Hit.java is able to convert the response to the java class for the parent perfectly fine, however when trying to convert its children, it fails because each InnerHit included with the parent Hit does not have a index specified in the response. I believe it was done under the intention that the the index of the InnerHits should always match the parent, however, the conversion is not aware of the Parent document's index for the InnerHit.

### How can one reproduce the bug?
_Steps to reproduce the behavior._
Reproduction steps would be to create two documents, and associate a parent-child relation to them. And then run a query for a property that would match the parent, and then adding the innerHits field:
```
it.innerHits {
            it.source {
                it.fetch(true)
            }
        }
```
### What is the expected behavior?
_A clear and concise description of what you expected to happen._
Hit.java should be able to convert the InnerHits into Hit objects properly.
### What is your host/environment?
_Operating system, version._
MacOS 14.2.1
### Do you have any screenshots?
_If applicable, add screenshots to help explain your problem._

### Do you have any additional context?
I have to wonder why the index is set to mandatory? I did see a discussion on elasticsearch back in 2016 https://github.com/elastic/elasticsearch/issues/18091 but I am not sure if it's the same.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -27,6 +27,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Fix partial success results for msearch_template ([#709](https://github.com/opensearch-project/opensearch-java/pull/709))
 - Fix deserialization of node stats response ([#745](https://github.com/opensearch-project/opensearch-java/pull/745))
 - Fix PutIndexTemplateRequest field deserialization ([#765](https://github.com/opensearch-project/opensearch-java/pull/765))
+- Fix InnerHits to no longer enforce the nullable Index field when converting to Hit. ([#825](https://github.com/opensearch-project/opensearch-java/issues/825))
 
 ### Security
 

--- a/java-client/src/main/java/org/opensearch/client/opensearch/core/search/Hit.java
+++ b/java-client/src/main/java/org/opensearch/client/opensearch/core/search/Hit.java
@@ -107,7 +107,7 @@ public class Hit<TDocument> implements JsonpSerializable {
 
     private Hit(Builder<TDocument> builder) {
 
-        this.index = ApiTypeHelper.requireNonNull(builder.index, this, "index");
+        this.index = builder.index;
         this.id = builder.id;
         this.score = builder.score;
         this.explanation = builder.explanation;
@@ -134,7 +134,7 @@ public class Hit<TDocument> implements JsonpSerializable {
     }
 
     /**
-     * Required - API name: {@code _index}
+     * API name: {@code _index}
      */
     public final String index() {
         return this.index;
@@ -281,8 +281,10 @@ public class Hit<TDocument> implements JsonpSerializable {
 
     protected void serializeInternal(JsonGenerator generator, JsonpMapper mapper) {
 
-        generator.writeKey("_index");
-        generator.write(this.index);
+        if (this.index != null) {
+            generator.writeKey("_index");
+            generator.write(this.index);
+        }
 
         if (this.id != null) {
             generator.writeKey("_id");
@@ -419,6 +421,8 @@ public class Hit<TDocument> implements JsonpSerializable {
      */
 
     public static class Builder<TDocument> extends ObjectBuilderBase implements ObjectBuilder<Hit<TDocument>> {
+
+        @Nullable
         private String index;
 
         @Nullable
@@ -478,7 +482,7 @@ public class Hit<TDocument> implements JsonpSerializable {
         /**
          * Required - API name: {@code _index}
          */
-        public final Builder<TDocument> index(String value) {
+        public final Builder<TDocument> index(@Nullable String value) {
             this.index = value;
             return this;
         }

--- a/java-client/src/test/java/org/opensearch/client/opensearch/core/search/InnerHitsResultTest.java
+++ b/java-client/src/test/java/org/opensearch/client/opensearch/core/search/InnerHitsResultTest.java
@@ -1,0 +1,86 @@
+package org.opensearch.client.opensearch.core.search;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+import java.io.StringReader;
+import org.junit.Test;
+import org.opensearch.client.json.JsonData;
+import org.opensearch.client.json.JsonpMapper;
+import org.opensearch.client.json.jsonb.JsonbJsonpMapper;
+
+public class InnerHitsResultTest {
+    private final JsonpMapper mapper = new JsonbJsonpMapper();
+    private final String storedSalary = "details.salary";
+    private final String storedJobId = "details.jobId";
+
+    /**
+     * test if the InnerHitsResult will build the Hit<JsonData>
+     */
+    @Test
+    public void testInnerHits() {
+
+        String classString = String.valueOf(hitResultWithIdIndex.innerHits().get("test_child").getClass());
+        assertEquals(classString, InnerHitsResult.class.toString());
+        // take hitResult and get the InnerHit
+        InnerHitsResult innerHitsResult = hitResultWithIdIndex.innerHits().get("test_child");
+        Hit<JsonData> innerHitResult = innerHitsResult.hits().hits().get(0);
+        assertNotNull(innerHitResult.index());
+        assertEquals(innerHitResult.index(), "_index");
+        assertNotNull(innerHitResult.id());
+        assertEquals(innerHitResult.id(), "child_id");
+    }
+
+    /**
+     * test if the InnerHitsResult will still build the Hit<JsonData> even if id and index is not specified
+     */
+    @Test
+    public void testInnerHitWithoutIdIndex() {
+
+        String classString = String.valueOf(hitResultNoIdIndex.innerHits().get("test_child").getClass());
+        assertEquals(classString, InnerHitsResult.class.toString());
+        // take hitResult and get the InnerHit
+        InnerHitsResult innerHitsResult = hitResultNoIdIndex.innerHits().get("test_child");
+        Hit<JsonData> innerHitResult = innerHitsResult.hits().hits().get(0);
+        // Id and index are now nullable.
+        assertNull(innerHitResult.index());
+        assertNull(innerHitResult.id());
+    }
+
+    private final String innerHitJsonWithNoIdOrIndex = "{\"key\":\"value\"}";
+    private final String innerHitJsonWithIdOrIndex = "{\"id\":\"value\",\"index\":\"value\"}";
+
+    private final Hit<JsonData> hitResultNoIdIndex = Hit.of(
+        it -> it.id("test_parent")
+            .index("_index")
+            .innerHits(
+                "test_child",
+                innerHitsResultBuilder -> innerHitsResultBuilder.hits(
+                    innerHitsMetadataBuilder -> innerHitsMetadataBuilder.total(total -> total.value(1).relation(TotalHitsRelation.Eq))
+                        .hits(
+                            innerHitsListMemberBuilder -> innerHitsListMemberBuilder.source(
+                                JsonData.from(mapper.jsonProvider().createParser(new StringReader(innerHitJsonWithNoIdOrIndex)), mapper)
+                            )
+                        )
+                )
+            )
+    );
+    private final Hit<JsonData> hitResultWithIdIndex = Hit.of(
+        it -> it.id("test_parent")
+            .index("_index")
+            .innerHits(
+                "test_child",
+                innerHitsResultBuilder -> innerHitsResultBuilder.hits(
+                    innerHitsMetadataBuilder -> innerHitsMetadataBuilder.total(total -> total.value(1).relation(TotalHitsRelation.Eq))
+                        .hits(
+                            innerHitsListMemberBuilder -> innerHitsListMemberBuilder.id("child_id")
+                                .index("_index")
+                                .source(
+                                    JsonData.from(mapper.jsonProvider().createParser(new StringReader(innerHitJsonWithIdOrIndex)), mapper)
+                                )
+                        )
+                )
+            )
+    );
+}

